cmd/kubectl-gadget/utils/trace.go

```diff
@@ -463,95 +466,157 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
+			// So, we do not add it to the map here.
+			nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
 		}
 
-		// We are only interested in Added and Modified event, as we want
-		// Status.State value to change.
-		// More particularly, we monitor Added event for gadget like dns and
-		// Modified for gadget like seccompadvisor.
-		if event.Type != watch.Added && event.Type != watch.Modified {
-			return false, nil
+		if trace.Status.OperationError != "" {
+			erroredTraces[trace.ObjectMeta.Name] = &traceList.Items[i]
+
+			continue
 		}
 
-		if event.Type == watch.Added {
-			// The API adds fake watch.Added events, so we use them to count the
-			// number of trace:
-			// To establish initial state, the watch begins with synthetic "Added"
-			// events of all resources instances that exist at the starting resource
-			// version
-			tracesNumber++
+		if !conditionFunction(&trace) {
+			continue
 		}
 
-		trace := event.Object.(*gadgetv1alpha1.Trace)
+		satisfiedTraces[trace.ObjectMeta.Name] = &traceList.Items[i]
+	}
 
-		if trace.Status.OperationWarning != "" {
-			nodeWarnings[trace.Spec.Node] = trace.Status.OperationWarning
+	tracesNumber := len(traceList.Items)
+
+	// We only watch the traces which did not satisfy conditionFunction.
+	tracesToWatch := tracesNumber - len(satisfiedTraces)
+	if tracesToWatch > 0 {
+		// We will need to watch events on them.
+		// For this, we will get a watcher on all the traces which share the same
+		// ID.
+		// Get a watcher on all the traces which have the same ID.
+		// Indeed, all the traces on different nodes but linked to one gadget share
+		// the same ID.
+		// We will also begin to monitor events since the above GET of the traces
+		// list.
+		watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
+		if err != nil {
+			return nil, err
 		}
 
-		if trace.Status.OperationError != "" {
-			watchedTracesNumber++
+		ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), traceTimeout)
+		_, err = untilWithoutRetry(ctx, watcher, func(event watch.Event) (bool, error) {
+			// This function will be executed until:
+			// 1. The number of watched traces equals the number of traces to watch,
+			// i.e. we dealt with the traces which interest us.
+			// 2. Or it returns an error.
+			// 3. Or time out is fired.
+			// NOTE In case 2 and 3, it exists, at least, one trace we did not deal
+			// with.
+			switch event.Type {
+			case watch.Deleted:
+				// If for some strange reasons (e.g. users deleted a trace during this
+				// operation) a trace is deleted, we need to take care of this by
+				// decrementing the tracesToWatch.
+				// Otherwise we would still wait for the old number and we would
+				// timeout.
+				tracesToWatch--
+
+				trace, ok := event.Object.(*gadgetv1alpha1.Trace)
+				if !ok {
+					return false, fmt.Errorf("Received object cannot be cast to Trace: %v", event.Object)
+				}
```

Comment:
Actually in this case the type assertion is not needed as you already checked event.Type and we're sure that for deleted events the object is a trace.